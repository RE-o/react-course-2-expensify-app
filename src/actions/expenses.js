import {getDatabase, off, ref, set, get, onValue, onChildRemoved, onChildChanged, onChildAdded, update, remove, push} from 'firebase/database';
import { app } from '../firebase/firebase';

const db = getDatabase(app);

// Actions needed
// ADD_EXPENSE. "... = {}" means that we set empty object as default
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    // here we need to return push... in order to use promise chaining (cfr. tests in actions/expenses)
    return push(ref(db, 'expenses'), expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({id} = {}) => {
  return (dispatch) => {
    return remove(ref(db, `expenses/${id}`)).then(() => {
      dispatch(removeExpense({id}));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// we need async action to fetch data
export const startSetExpenses = () => {
  return (dispatch) => {
    // get data and then parse it into an array
    return get(ref(db, 'expenses')).then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};
