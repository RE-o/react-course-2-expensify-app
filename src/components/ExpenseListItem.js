import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import 'numeral/locales/it';
numeral.locale('it');

export const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>Title: { description }</h3>
    </Link>
    <p>
      {numeral(amount / 100).format('$0,0.00')}
      -
      {moment(createdAt).format('Do MMMM, YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;
