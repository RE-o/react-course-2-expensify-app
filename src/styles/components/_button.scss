.button {
  color: white;
  background: $blue;
  border: none;
  display: inline-block; // to style also simple Links as buttons. By default links are displayed inline whence no margin and padding 
  font-size: $font-size-large;
  font-weight: 300;
  line-height: 1; // the button text is on 1 line
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--secondary {
  background: #888;
}
