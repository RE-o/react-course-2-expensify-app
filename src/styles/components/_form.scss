.form {
  display: flex;
  flex-direction: column;
  > * { // all direct descendant
    margin-bottom: $m-size;
  }
}

.form__error {
  font-style: italic;
  margin: 0 0 $m-size 0;
}
