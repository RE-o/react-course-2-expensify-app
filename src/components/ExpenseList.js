import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from '../components/ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// we exported it since we snapshot test it unconnected
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
      props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No expenses</span>
        </div>
      ) : (
          props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
        )
    }
    </div>
    {/* in order to not pass down the id use the rest operator */}
    {/* { props.expenses.map(({id, ...expense}) => <ExpenseListItem key={id} {...expense} />) } */}
  </div>
);

const mapStateToProps = (state) => {
  const selected = selectExpenses(state.expenses, state.filters);

  return {
    expenses: selected
  };
};

export default connect(mapStateToProps)(ExpenseList);
