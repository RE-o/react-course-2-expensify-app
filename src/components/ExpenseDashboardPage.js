import React from 'react';
import ExpenseList from '../components/ExpenseList';
import ExpenseListFilters from '../components/ExpenseListFilters';
import ExpensesSummary from '../components/ExpensesSummary';

export default () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

