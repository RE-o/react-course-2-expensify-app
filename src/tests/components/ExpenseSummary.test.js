import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should correctly render ExpensesSummary with 1 expese', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={12020} />)
  expect(wrapper).toMatchSnapshot();
});

test('should render 2 expensis correctly with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={12020} />)

  expect(wrapper).toMatchSnapshot();
});
