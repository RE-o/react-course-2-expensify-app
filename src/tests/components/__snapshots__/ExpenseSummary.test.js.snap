// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expese 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling 
    $120.20
  </h1>
</div>
`;

exports[`should render 2 expensis correctly with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling 
    $120.20
  </h1>
</div>
`;
